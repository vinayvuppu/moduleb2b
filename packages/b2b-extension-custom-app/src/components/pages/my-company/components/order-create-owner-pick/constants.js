// eslint-disable-next-line
export const OWNER_TYPES = {
  OWNER_ME: 'me',
  OWNER_EMPLOYEE: 'employee',
  OWNER_COMPANY: 'company',
};
