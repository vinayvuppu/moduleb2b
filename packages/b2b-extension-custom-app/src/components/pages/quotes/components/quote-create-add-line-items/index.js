export { default } from './quote-create-add-line-items';
