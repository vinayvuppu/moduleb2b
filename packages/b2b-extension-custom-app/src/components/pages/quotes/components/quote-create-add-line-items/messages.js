import { defineMessages } from 'react-intl';

export default defineMessages({
  title: {
    id: 'Quote.Create.Step.LineItems.title',
    defaultMessage: 'Add item',
  },
  searchPlaceHolder: {
    id: 'Quote.Create.Step.LineItems.customsearch.placeHolder',
    defaultMessage: 'Search by term.',
  },
  addVariantSuccess: {
    id: 'Quote.Create.Step.LineItems.search.addVariantSuccess',
    defaultMessage: 'Variant {sku} has been added successfully to the quote',
  },
  addVariantFailure: {
    id: 'Quote.Create.Step.LineItems.search.addVariantFailure',
    description:
      'Message when the variant has not been added to the cart due to an error',
    defaultMessage:
      'The variant that you are trying to add is invalid. Price format does not match the selected currency',
  },
  addVariantFailureTaxRate: {
    id: 'Quote.Create.Step.LineItems.search.addVariantFailureTaxRate',
    description:
      'Message when the variant has not been added to the cart due to an error with the tax rate',
    defaultMessage:
      'The variant that you are trying to add is invalid because its tax category does not have a rate defined for the shipping country of the quote',
  },
});
