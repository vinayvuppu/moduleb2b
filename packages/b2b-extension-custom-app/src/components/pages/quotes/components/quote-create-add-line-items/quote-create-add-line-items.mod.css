.panel {
  width: 100%;
}

.products-container {
  display: flex;
  justify-content: space-around;
  flex-wrap: wrap;
}
