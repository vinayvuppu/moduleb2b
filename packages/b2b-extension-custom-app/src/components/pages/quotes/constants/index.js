// eslint-disable-next-line
export const QUOTE_TYPES = {
  INITIAL: 'initial',
  SUBMITTED: 'submitted',
  APPROVED: 'approved',
  CLOSED: 'closed',
  DECLINED: 'declined',
  EXPIRED: 'expired',
  PLACED: 'placed',
};

export const QUOTE_DETAILS_TAB_NAMES = {
  GENERAL: 'general',
  COMMENTS: 'comments',
};
