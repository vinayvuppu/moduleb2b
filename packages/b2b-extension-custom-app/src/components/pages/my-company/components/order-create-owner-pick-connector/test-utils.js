/* eslint-disable import/prefer-default-export */
import { GRAPHQL_TARGETS } from '@commercetools-frontend/constants';
import flatMap from 'lodash.flatmap';
import intersection from 'lodash.intersection';
import FetchCustomersQuery from './order-create-customer-pick-connector.graphql';

export const createCustomersFetchQueryMock = (stores = []) => {
  const whereClause =
    stores.length === 0
      ? undefined
      : `(stores(key = "${stores.join(',')}") or stores is empty)`;
  const customers = [
    {
      id: 'customer-1',
      version: 4,
      firstName: 'Anna',
      lastName: 'Banana',
      email: 'annabanana@blablub.com',
      customerNumber: '111',
      companyName: null,
      customerGroup: null,
      stores: [
        {
          id: 'store-2',
          key: 'usa',
          nameAllLocales: [{ locale: 'en-GB', value: 'USA' }],
        },
      ],
      addresses: [
        {
          id: '_ZqTe0es',
          firstName: '3333',
          streetName: 'Fakestreet',
          streetNumber: '23',
          country: 'DE',
          additionalAddressInfo: 'Additional address info2',
          additionalStreetInfo: 'Additional street info',
          email: null,
          phone: null,
          city: null,
          company: null,
          lastName: null,
          postalCode: null,
          region: null,
        },
      ],
    },
    {
      id: 'customer-2',
      version: 1,
      firstName: 'Ali',
      lastName: 'Baba',
      customerNumber: '235',
      email: 'alibaba@testest.de',
      companyName: null,
      customerGroup: null,
      stores: [
        {
          id: 'store-1',
          key: 'germany',
          nameAllLocales: [{ locale: 'en-GB', value: 'Germany' }],
        },
      ],
      addresses: [],
    },
    {
      id: 'customer-3',
      version: 3,
      customerNumber: '128',
      firstName: 'Amin',
      lastName: 'Rocks',
      email: 'email@ct.de',
      companyName: null,
      customerGroup: null,
      stores: [],
      addresses: [
        {
          id: 'yUQGmqRG',
          firstName: 'first',
          lastName: 'last2',
          company: 'company',
          streetName: 'Street name',
          streetNumber: 'House number',
          city: 'city',
          postalCode: '66215',
          region: 'region',
          country: 'US',
          additionalAddressInfo: 'Additional address info2',
          additionalStreetInfo: 'Additional street info',
          email: 'email2@email.com',
          phone: 'phone',
        },
      ],
    },
  ].filter(customer => {
    // if we don't pass stores then we don't filter
    if (!stores) return true;

    const customerStoresArray = flatMap(customer.stores, s => s.key);
    return (
      // true if the customer belongs to one of the stores passed
      intersection(customerStoresArray, stores).length !== 0 ||
      // or the customer belongs to no store
      customerStoresArray.length === 0
    );
  });

  return {
    request: {
      variables: {
        target: GRAPHQL_TARGETS.COMMERCETOOLS_PLATFORM,
        projectKey: 'context-playground',
        limit: 20,
        locale: 'en',
        offset: 0,
        sort: ['createdAt desc'],
        where: whereClause,
      },
      query: FetchCustomersQuery,
    },
    result: {
      data: {
        customers: {
          count: 3,
          offset: 0,
          total: 3,
          results: customers,
        },
      },
    },
  };
};
