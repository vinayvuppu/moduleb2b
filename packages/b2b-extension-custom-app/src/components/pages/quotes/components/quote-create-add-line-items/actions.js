import { actions as sdkActions } from '@commercetools-frontend/sdk';
import { MC_API_PROXY_TARGETS } from '@commercetools-frontend/constants';

// eslint-disable-next-line
export const fetchProducts = options =>
  sdkActions.get({
    mcApiProxyTarget: MC_API_PROXY_TARGETS.COMMERCETOOLS_PLATFORM,
    service: 'productProjectionsSearch',
    options: { ...options, expand: ['productType'] },
  });
