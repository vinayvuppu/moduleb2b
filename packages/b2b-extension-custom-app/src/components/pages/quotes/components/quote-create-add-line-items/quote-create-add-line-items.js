import React from 'react';
import PropTypes from 'prop-types';
import { compose } from 'recompose';
import { connect } from 'react-redux';
import { FormattedMessage, injectIntl } from 'react-intl';

import { DOMAINS } from '@commercetools-frontend/constants';
import * as globalActions from '@commercetools-frontend/actions-global';
import { Spacings, Text, LoadingSpinner } from '@commercetools-frontend/ui-kit';
import Pagination from '@commercetools-local/core/components/search/pagination';
import {
  ProjectExtensionProviderForImageRegex,
  withApplicationContext,
} from '@commercetools-frontend/application-shell-connectors';
import SearchInput from '@commercetools-local/core/components/fields/search-input';

import styles from './quote-create-add-line-items.mod.css';
import messages from './messages';

import OrderCreateProductsListItem from '../../../my-company/components/order-create-products-list-item';

import * as productActions from './actions';

export class QuoteCreateAddLineItems extends React.PureComponent {
  static displayName = 'QuoteCreateAddLineItems';

  state = {
    products: null,
    facets: null,
    total: 0,
    perPage: 20,
    page: 1,
    searchTerm: '',
  };

  static propTypes = {
    renderSaveToolbarStep: PropTypes.func.isRequired,
    addLineItem: PropTypes.func.isRequired,
    quote: PropTypes.object,
    company: PropTypes.shape({
      customerGroup: PropTypes.shape({
        id: PropTypes.string.isRequired,
      }),
    }),
    projectKey: PropTypes.string.isRequired,
    allLineItems: PropTypes.array.isRequired,
    showNotification: PropTypes.func.isRequired,
    onActionError: PropTypes.func.isRequired,
    fetchProducts: PropTypes.func.isRequired,
    // HoC
    intl: PropTypes.shape({ formatMessage: PropTypes.func.isRequired })
      .isRequired,
    // withApplicationContext
    dataLocale: PropTypes.string.isRequired,
  };

  handleAddVariantToCart = variant => {
    return this.props
      .addLineItem({ sku: variant.sku, quantity: parseFloat(variant.quantity) })
      .then(
        () => {
          this.props.showNotification({
            kind: 'success',
            domain: DOMAINS.SIDE,
            text: this.props.intl.formatMessage(messages.addVariantSuccess, {
              sku: variant.sku,
            }),
          });
        },
        graphQLErrors => {
          if (graphQLErrors.length > 0) {
            const graphQLError = graphQLErrors[0];
            let errorMessage = graphQLError.message;
            if (graphQLError.code === 'MatchingPriceNotFound')
              errorMessage = this.props.intl.formatMessage(
                messages.addVariantFailure
              );
            if (graphQLError.code === 'MissingTaxRateForCountry')
              errorMessage = this.props.intl.formatMessage(
                messages.addVariantFailureTaxRate
              );
            this.props.showNotification({
              kind: 'error',
              domain: DOMAINS.SIDE,
              text: errorMessage,
            });
          } else {
            this.props.onActionError(
              graphQLErrors,
              'OrderCreateVariantSearch/updateCart'
            );
          }
        }
      );
  };

  setSearchTerm = nextSearchTerm =>
    this.setState(prevState => ({
      ...prevState,
      searchTerm: nextSearchTerm,
    }));

  handlePageChange = nextPage =>
    this.setState(prevState => ({
      ...prevState,
      page: nextPage,
    }));

  handlePerPageChange = nextPerPage =>
    this.setState(prevState => ({
      ...prevState,
      perPage: nextPerPage,
    }));

  searchProducts = async () => {
    try {
      const options = {
        filter: [`variants.scopedPrice:exists`],
        page: this.state.page,
        perPage: this.state.perPage,
        priceCurrency: this.props.quote.totalPrice.currencyCode,
        priceCustomerGroup: this.props.company?.customerGroup?.id,
        ...(this.state.searchTerm && {
          text: {
            value: this.state.searchTerm,
            language: this.props.dataLocale,
          },
        }),
        ...(!this.state.searchTerm && {
          sort: [{ by: 'id', direction: 'asc' }], // this is done for mantain order when the searchTerm is not provided
        }),
      };

      const response = await this.props.fetchProducts(options);

      this.setState(prevState => ({
        ...prevState,
        products: response.results,
        total: response.total,
      }));
    } catch (error) {
      this.props.onActionError(error, 'ProductPage/fetchProducts');
    }
  };

  async componentDidMount() {
    await this.searchProducts();
  }

  async componentDidUpdate(prevProps, prevState) {
    if (
      prevState.searchTerm !== this.state.searchTerm ||
      prevState.page !== this.state.page ||
      prevState.perPage !== this.state.perPage
    ) {
      await this.searchProducts();
    }
  }

  render() {
    if (!this.state.products) {
      return <LoadingSpinner />;
    }

    return (
      <Spacings.Stack scale="m">
        <Spacings.Inline scale="l" justifyContent="space-between">
          <Text.Headline as="h2">
            <FormattedMessage {...messages.title} />
          </Text.Headline>
        </Spacings.Inline>
        <SearchInput
          initialValue={this.state.searchTerm}
          onSubmit={this.setSearchTerm}
          placeholder={this.props.intl.formatMessage(
            messages.searchPlaceHolder
          )}
        />
        <ProjectExtensionProviderForImageRegex>
          <div className={styles['products-container']}>
            {this.state.products.map(product => (
              <OrderCreateProductsListItem
                {...product}
                key={product.id}
                onAddVariantToCart={this.handleAddVariantToCart}
                isAddingVariant={false /* this.props.cartUpdater.isLoading */}
              />
            ))}
          </div>
        </ProjectExtensionProviderForImageRegex>
        <Pagination
          totalItems={this.state.total}
          perPage={this.state.perPage}
          page={this.state.page}
          onPerPageChange={this.handlePerPageChange}
          onPageChange={this.handlePageChange}
        />
        {this.props.renderSaveToolbarStep()}
      </Spacings.Stack>
    );
  }
}

const mapStateToProps = (state, ownProps) => {
  return {
    allLineItems: ownProps.quote?.lineItems || [],
  };
};

const mapDispatchToProps = {
  showNotification: globalActions.showNotification,
  onActionError: globalActions.handleActionError,
  fetchProducts: productActions.fetchProducts,
};

export default compose(
  withApplicationContext(applicationContext => ({
    dataLocale: applicationContext.dataLocale,
  })),
  injectIntl,
  connect(mapStateToProps, mapDispatchToProps)
)(QuoteCreateAddLineItems);
