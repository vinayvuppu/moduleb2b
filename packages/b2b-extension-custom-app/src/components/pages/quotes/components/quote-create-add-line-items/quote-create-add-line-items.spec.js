import React from 'react';
import { shallow } from 'enzyme';
import { DOMAINS } from '@commercetools-frontend/constants';
import { intlMock } from '@commercetools-local/test-utils';
import { QuoteCreateAddLineItems } from './quote-create-add-line-items';

const lineItems = [
  {
    id: 'line-item-id',
    productId: 'product-id',
    nameAllLocales: [
      {
        locale: 'en',
        value: 'Top Jucca black',
      },
    ],
    variant: {
      sku: 'M0E20000000EC8N',
      images: [
        {
          url: 'url-1',
        },
      ],
    },
    price: {
      id: 'price-id-1',
      value: {
        currencyCode: 'USD',
        centAmount: 100,
      },
    },
    quantity: 1,
    totalPrice: {
      currencyCode: 'USD',
      centAmount: 100,
    },
  },
];
const createTestProps = props => ({
  renderSaveToolbarStep: jest.fn(),
  allLineItems: lineItems,
  company: {
    customerGroup: {
      id: 'customer-group-id',
    },
  },
  quote: {
    id: 'quote-id',
    company: {
      id: 'company-id',
      name: 'company name',
    },
    customerGroup: {
      id: 'customer-group-id',
    },
    employeeEmail: 'employee@company.com',
    totalPrice: {
      currencyCode: 'USD',
      centAmount: 100,
    },
    lineItems,
  },
  addLineItem: jest.fn(),
  showNotification: jest.fn(),
  onActionError: jest.fn(),
  fetchProducts: jest.fn(),
  projectKey: 'test-project',
  intl: intlMock,
  // withApplicationContext
  dataLocale: 'en',
  ...props,
});

describe('render', () => {
  describe('when loading', () => {
    let props;
    let wrapper;
    beforeEach(() => {
      props = createTestProps({ fetchProducts: jest.fn() });
      wrapper = shallow(<QuoteCreateAddLineItems {...props} />);
    });

    it('should match snapshot', () => {
      expect(wrapper).toMatchSnapshot();
    });

    it('should render the LoadingSpinner', () => {
      expect(wrapper).toRender('LoadingSpinner');
    });
  });
  describe('when the quote is empty', () => {
    let props;
    let wrapper;
    beforeEach(() => {
      props = createTestProps({ allLineItems: [] });
      wrapper = shallow(<QuoteCreateAddLineItems {...props} />);
      wrapper.setState({ totals: 1, products: [{ id: 'product-id-1' }] });
    });

    it('should match snapshot', () => {
      expect(wrapper).toMatchSnapshot();
    });
    it('should render the title', () => {
      expect(wrapper).toRender('TextHeadline');
    });
  });
});

describe('callbacks', () => {
  describe('when adding a variant to the cart successfully', () => {
    let props;
    let wrapper;

    beforeEach(async () => {
      props = createTestProps({
        addLineItem: jest.fn().mockResolvedValue(),
      });
      wrapper = shallow(<QuoteCreateAddLineItems {...props} />);
      await wrapper.instance().handleAddVariantToCart({
        sku: '123',
        quantity: 1,
      });
    });
    it('should call `addLineItem`', () => {
      expect(props.addLineItem).toHaveBeenCalledTimes(1);
    });
    it('should call showNotification', () => {
      expect(props.showNotification).toHaveBeenCalledTimes(1);
    });
    it('should call showNotification with parameters', () => {
      expect(props.showNotification).toHaveBeenCalledWith({
        kind: 'success',
        domain: DOMAINS.SIDE,
        text: 'Quote.Create.Step.LineItems.search.addVariantSuccess',
      });
    });
  });

  describe('when there is an error with adding a variant to the cart', () => {
    let props;
    let wrapper;

    beforeEach(async () => {
      props = createTestProps({
        addLineItem: jest
          .fn()
          .mockRejectedValue(new Error('Something went wrong')),
      });
      wrapper = shallow(<QuoteCreateAddLineItems {...props} />);
      await wrapper
        .instance()
        .handleAddVariantToCart({
          sku: '123',
          quantity: 1,
        })
        .then(
          () => {},
          () => {}
        );
    });
    it('should call onActionError', () => {
      expect(props.onActionError).toHaveBeenCalledTimes(1);
    });
  });

  describe('setSearchTerm', () => {
    let wrapper;
    let props;
    beforeEach(() => {
      props = createTestProps();
      wrapper = shallow(<QuoteCreateAddLineItems {...props} />);
      wrapper = wrapper.setState({
        totals: 1,
        products: [{ id: 'product-id-1' }],
      });
      wrapper.instance().setSearchTerm('testing');
    });

    it('should change the state', () => {
      expect(wrapper).toHaveState('searchTerm', 'testing');
    });
  });

  describe('handlePagechange', () => {
    let wrapper;
    let props;
    beforeEach(() => {
      props = createTestProps();
      wrapper = shallow(<QuoteCreateAddLineItems {...props} />);
      wrapper = wrapper.setState({
        totals: 1,
        products: [{ id: 'product-id-1' }],
      });
      wrapper.instance().handlePageChange(10);
    });

    it('should change the state', () => {
      expect(wrapper).toHaveState('page', 10);
    });
  });

  describe('handlePerPageChange', () => {
    let wrapper;
    let props;
    beforeEach(() => {
      props = createTestProps();
      wrapper = shallow(<QuoteCreateAddLineItems {...props} />);
      wrapper = wrapper.setState({
        totals: 1,
        products: [{ id: 'product-id-1' }],
      });
      wrapper.instance().handlePerPageChange(10);
    });

    it('should change the state', () => {
      expect(wrapper).toHaveState('perPage', 10);
    });
  });

  describe('searchProducts', () => {
    let wrapper;
    let props;

    describe('when there is not a search term', () => {
      beforeEach(async () => {
        props = createTestProps();
        wrapper = shallow(<QuoteCreateAddLineItems {...props} />);
        wrapper = wrapper.setState({
          totals: 1,
          products: [{ id: 'product-id-1' }],
        });
        await wrapper.instance().searchProducts();
      });

      it('should call to fetchProducts with sort id', () => {
        expect(props.fetchProducts).toHaveBeenCalledWith({
          filter: ['variants.scopedPrice:exists'],
          page: 1,
          perPage: 20,
          priceCurrency: 'USD',
          priceCustomerGroup: 'customer-group-id',
          sort: [{ by: 'id', direction: 'asc' }],
        });
      });
    });
    describe('when there is a search term', () => {
      beforeEach(async () => {
        props = createTestProps();
        wrapper = shallow(<QuoteCreateAddLineItems {...props} />);
        wrapper = wrapper.setState({
          searchTerm: 'testing',
          totals: 1,
          products: [{ id: 'product-id-1' }],
        });
        await wrapper.instance().searchProducts();
      });

      it('should call to fetchProducts with text filter', () => {
        expect(props.fetchProducts).toHaveBeenCalledWith({
          filter: ['variants.scopedPrice:exists'],
          page: 1,
          perPage: 20,
          priceCurrency: 'USD',
          priceCustomerGroup: 'customer-group-id',
          text: { language: 'en', value: 'testing' },
        });
      });
    });
  });
});
