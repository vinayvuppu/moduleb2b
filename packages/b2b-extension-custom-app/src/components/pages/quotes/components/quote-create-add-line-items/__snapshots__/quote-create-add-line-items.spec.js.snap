// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render when loading should match snapshot 1`] = `
<LoadingSpinner
  scale="l"
/>
`;

exports[`render when the quote is empty should match snapshot 1`] = `
<Stack
  alignItems="stretch"
  scale="m"
>
  <Inline
    alignItems="flex-start"
    justifyContent="space-between"
    scale="l"
  >
    <TextHeadline
      as="h2"
    >
      <FormattedMessage
        defaultMessage="Add item"
        id="Quote.Create.Step.LineItems.title"
        values={Object {}}
      />
    </TextHeadline>
  </Inline>
  <SearchInput
    initialValue=""
    onSubmit={[Function]}
    placeholder="Quote.Create.Step.LineItems.customsearch.placeHolder"
  />
  <ProjectExtensionProviderForImageRegex>
    <div
      className="products-container"
    >
      <OrderCreateProductsListItem
        id="product-id-1"
        isAddingVariant={false}
        key="product-id-1"
        onAddVariantToCart={[Function]}
      />
    </div>
  </ProjectExtensionProviderForImageRegex>
  <Pagination
    onPageChange={[Function]}
    onPerPageChange={[Function]}
    page={1}
    pageSizes={
      Array [
        20,
        50,
      ]
    }
    perPage={20}
    totalItems={0}
  />
</Stack>
`;
